### PR TITLE
Hsm/implement and test sha3-512 algorithm

### PR DESCRIPTION
**Pull request by YochevedSalzer**
Overview: This PR adds the SHA3-512 hashing algorithm to our project. It includes the implementation and corresponding unit tests.

Changes:

Implemented SHA3-512 in src/SHA3-512.cpp.
Added tests in tests/SHA3-512_tests.cpp.
Updated CMakeLists.txt for build configuration.
Benefits:

Enhanced security with SHA3-512.
Comprehensive test coverage.
Please review the implementation and tests.

### CODE REVIEW

File: src/SHA3-512.cpp [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    init(); // Initialize the state
+}
+
+// Initialize or reset the context
+void SHA3_512::init()
+{
+    std::memset(S, 0, sizeof(S)); // Clear state
+    std::memset(buffer, 0, sizeof(buffer)); // Clear buffer
+    buffer_length = 0; // Reset buffer length
+
```
Comment (TobyRoth) [inline src/SHA3-512.cpp]: Remove unnecessary blank line

File: src/SHA3-512.cpp [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    std::size_t add_last = in_len % 72; // Remaining bytes
+    if (72 - add_last == 1) {
+        input[in_len] = 0x86; // Padding byte for the case where only one byte is left
+    } else {
+        input[in_len] = 0x06; // Padding byte for general case
+        for (int i = 1; i < (71 - add_last); i++)
+            input[in_len + i] = 0x00; // Fill with zeroes
+        input[in_len + (71 - add_last)] = 0x80; // Final padding byte
+        in_len += (72 - add_last); // Update length
+    }
```
Comment (TobyRoth) [inline src/SHA3-512.cpp:158]: 1. Consider using a const or macro instead of a magic number 72
2. Add a blank line after paragraph

File: src/SHA3-512.cpp [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+            // Apply the Keccak-f function to update the state
+            f_function(S);
+
+            // Reset buffer length for the next block of data
+            buffer_length = 0;
+        }
+    }
+}
+
+
```
Comment (TobyRoth) [inline src/SHA3-512.cpp:335]: Remove unnecessary blank line (only one is needed)

File: src/SHA3-512.cpp [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    return convertToString(S);
+}
+
+// Compute the hash of the input data in one step
+std::string SHA3_512::compute(const std::vector<uint8_t>& input)
+{
+    init();
+    update(input.data(), input.size());
+    return finalize();
+}
```
Comment (TobyRoth) [inline src/SHA3-512.cpp:420]: Please add error handling and logging messages in each possible place.

File: .vscode/settings.json
Comment (TobyRoth) [inline .vscode/settings.json:1]: Should be removed

File: .vscode/tasks.json
Comment (TobyRoth) [inline .vscode/tasks.json:1]: Also this file should not be part of the commit

File: src/SHA3-512.cpp [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+        in_len += (BLOCK_SIZE - add_last); // Update length
+    } else {
+        sha3_512logger.logMessage(logger::LogLevel::ERROR, "Unexpected padding error");
+        throw std::runtime_error("Padding failed due to unexpected error. Remaining bytes: " + std::to_string(add_last));
+    }
+
+    absorb_times += 1; // Include the final padding block
+}
+
+
```
Comment (KamaTechOrg) [inline src/SHA3-512.cpp]: No need for 2 blank line between the functions, do the same for all the places in the code

File: tests/SHA3-512_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+//     std::vector<std::vector<uint8_t>> data = {std::vector<uint8_t>(specialCharsString.begin(), specialCharsString.end())};
+//     std::string expectedHash = "4d977d5cf11054da4e40e28c7ea79fe6722aa11b0338cbacd92830cebee50787ab5b9e28adb25250a44c9933d84b610ea6bb7e992bdabcd3d45e56e996d1410c";
+//     sha3_512_test(data, expectedHash);
+// }
```
Comment (KamaTechOrg) [inline tests/SHA3-512_tests.cpp]: Is it not not necessary anymore?

Comment (YochevedSalzer) [reply tests/SHA3-512_tests.cpp]: I wanted to try testing a very long message and it didnt work when i did it, so i left it back then.
Now i wanted to check it out properly but because I did rebase to hsm-main I'm getting errors from other algorithms there. Am I supposed to fix those errors even though its not actually my branch? Or should i just drop it?
